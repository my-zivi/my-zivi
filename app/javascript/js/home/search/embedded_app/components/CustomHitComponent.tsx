import { ComponentType } from 'preact';
import { InfiniteHitsProvided } from 'react-instantsearch-core';
import React from 'preact/compat';
import { connectInfiniteHits } from 'react-instantsearch-dom';
import { isEmpty } from 'lodash';
import { JobPostingSearchHit } from 'js/home/search/embedded_app/types';
import JobPosting from './JobPosting';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const noResultsImage = require('../../../../../images/home/no-results.svg');

const NoResults: React.FunctionComponent = () => (
  <div className="d-flex flex-column align-items-center w-100 mt-4">
    <img src={noResultsImage} alt={MyZivi.translations.search.noResultsFound} className="w-25" />
    <h3 className="text-muted mt-4">{MyZivi.translations.search.noResultsFound}</h3>
  </div>
);

const renderHits = (hits: JobPostingSearchHit[]) => (
  hits.map((hit) => (
    <div className="col-12 col-xl-6 mb-4">
      <JobPosting hit={hit} key={hit.objectID} />
    </div>
  ))
);

const Hits: ComponentType<InfiniteHitsProvided> = ({ hits, hasMore, refineNext }) => (
  <div className="row">
    {isEmpty(hits) ? <NoResults /> : renderHits(hits)}
    {hasMore && (
      <div className="col-12 text-center mt-3">
        <button className="btn btn-link" onClick={refineNext}>{MyZivi.translations.search.loadMore}</button>
      </div>
    )}
  </div>
);

export default connectInfiniteHits(Hits);
