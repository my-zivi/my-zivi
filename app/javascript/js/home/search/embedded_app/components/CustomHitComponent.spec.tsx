import { mount } from 'enzyme';
import React from 'preact/compat';
import { InstantSearch } from 'react-instantsearch-dom';
import { defer } from 'lodash';
import CustomHitComponent from './CustomHitComponent';
import { mockClient } from '../../../../tests/algolia/mocks';

describe('CustomHitComponent', () => {
  it('renders correctly', () => {
    const rendered = mount(
      <InstantSearch searchClient={mockClient} indexName="JobPosting">
        <CustomHitComponent />
      </InstantSearch>,
    );

    return new Promise((resolve) => {
      defer(() => {
        rendered.update();
        const hitsComponent = rendered.find('Hits');
        expect(hitsComponent).toMatchSnapshot();
        resolve(undefined);
      });
    });
  });

  describe('when no results have been loaded yet', () => {
    it('renders no results page', () => {
      const rendered = mount(
        <InstantSearch searchClient={mockClient} indexName="JobPosting">
          <CustomHitComponent />
        </InstantSearch>,
      );

      expect(rendered.find('Hits')).toMatchSnapshot();
    });
  });
});
