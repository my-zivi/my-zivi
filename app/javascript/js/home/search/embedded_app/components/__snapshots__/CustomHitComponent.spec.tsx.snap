// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomHitComponent renders correctly 1`] = `
<Hits
  createURL={[Function]}
  hasMore={false}
  hasPrevious={false}
  hits={
    Array [
      Object {
        "__position": 1,
        "_highlightResult": Object {
          "organization_display_name": Object {
            "fullyHighlighted": false,
            "matchLevel": "full",
            "matchedWords": Array [
              "al",
            ],
            "value": "<ais-highlight-0000000000>Al</ais-highlight-0000000000>",
          },
          "plain_description": Object {
            "fullyHighlighted": false,
            "matchLevel": "full",
            "matchedWords": Array [
              "al",
            ],
            "value": "100% Dauer <ais-highlight-0000000000>Al</ais-highlight-0000000000>psommer",
          },
          "title": Object {
            "fullyHighlighted": false,
            "matchLevel": "full",
            "matchedWords": Array [
              "al",
            ],
            "value": "<ais-highlight-0000000000>Al</ais-highlight-0000000000>p-Pflege",
          },
        },
        "brief_description": "Bereich Kulturlandschaft, Biodiversität und Landschaftsqualität",
        "canton_display_name": "Obwalden",
        "category_display_name": "Landwirtschaft",
        "contact_information": "Dieser Betrieb ist noch nicht bei MyZivi registriert. Bitte bewerbe Dich im EZIVI.",
        "description": "MyDescription <b>Using HTML</b>",
        "featured_as_new": true,
        "icon_url": "/myzivi-logo.jpg",
        "identification_number": 89372,
        "language": "german",
        "link": "http://localhost:3000/job_postings/3",
        "minimum_service_months": 1,
        "objectID": "3",
        "organization_display_name": "Alpbetrieb Untersteiglen-Schwand im Tal-Melchsee-Frutt",
        "plain_description": "100% Dauer Alpsommer: von 01.05. bis 15.10",
        "preferred_skills": null,
        "priority_program": true,
        "publication_date": "2021-06-14",
        "published": true,
        "relevancy": 1,
        "required_skills": "<p> Motivation und physische Kondition für harte körperliche Arbeit im Freien</p>",
        "sub_category_display_name": null,
        "title": "Alp-Pflege",
      },
    ]
  }
  refine={[Function]}
  refineNext={[Function]}
  refinePrevious={[Function]}
>
  <div
    className="row"
  >
    <div
      className="col-12 col-xl-6 mb-4"
    >
      <ConnectorWrapper
        hit={
          Object {
            "__position": 1,
            "_highlightResult": Object {
              "organization_display_name": Object {
                "fullyHighlighted": false,
                "matchLevel": "full",
                "matchedWords": Array [
                  "al",
                ],
                "value": "<ais-highlight-0000000000>Al</ais-highlight-0000000000>",
              },
              "plain_description": Object {
                "fullyHighlighted": false,
                "matchLevel": "full",
                "matchedWords": Array [
                  "al",
                ],
                "value": "100% Dauer <ais-highlight-0000000000>Al</ais-highlight-0000000000>psommer",
              },
              "title": Object {
                "fullyHighlighted": false,
                "matchLevel": "full",
                "matchedWords": Array [
                  "al",
                ],
                "value": "<ais-highlight-0000000000>Al</ais-highlight-0000000000>p-Pflege",
              },
            },
            "brief_description": "Bereich Kulturlandschaft, Biodiversität und Landschaftsqualität",
            "canton_display_name": "Obwalden",
            "category_display_name": "Landwirtschaft",
            "contact_information": "Dieser Betrieb ist noch nicht bei MyZivi registriert. Bitte bewerbe Dich im EZIVI.",
            "description": "MyDescription <b>Using HTML</b>",
            "featured_as_new": true,
            "icon_url": "/myzivi-logo.jpg",
            "identification_number": 89372,
            "language": "german",
            "link": "http://localhost:3000/job_postings/3",
            "minimum_service_months": 1,
            "objectID": "3",
            "organization_display_name": "Alpbetrieb Untersteiglen-Schwand im Tal-Melchsee-Frutt",
            "plain_description": "100% Dauer Alpsommer: von 01.05. bis 15.10",
            "preferred_skills": null,
            "priority_program": true,
            "publication_date": "2021-06-14",
            "published": true,
            "relevancy": 1,
            "required_skills": "<p> Motivation und physische Kondition für harte körperliche Arbeit im Freien</p>",
            "sub_category_display_name": null,
            "title": "Alp-Pflege",
          }
        }
        key="3"
      >
        <Consumer>
          <Consumer>
            <AlgoliaInsights(JobPosting)
              contextValue={
                Object {
                  "createHrefForState": [Function],
                  "mainTargetedIndex": "JobPosting",
                  "onInternalStateUpdate": [Function],
                  "onSearchForFacetValues": [Function],
                  "onSearchParameters": [Function],
                  "onSearchStateChange": [Function],
                  "store": Object {
                    "getState": [Function],
                    "setState": [Function],
                    "subscribe": [Function],
                  },
                  "widgetsManager": Object {
                    "getWidgets": [Function],
                    "registerWidget": [Function],
                    "update": [Function],
                  },
                }
              }
              hit={
                Object {
                  "__position": 1,
                  "_highlightResult": Object {
                    "organization_display_name": Object {
                      "fullyHighlighted": false,
                      "matchLevel": "full",
                      "matchedWords": Array [
                        "al",
                      ],
                      "value": "<ais-highlight-0000000000>Al</ais-highlight-0000000000>",
                    },
                    "plain_description": Object {
                      "fullyHighlighted": false,
                      "matchLevel": "full",
                      "matchedWords": Array [
                        "al",
                      ],
                      "value": "100% Dauer <ais-highlight-0000000000>Al</ais-highlight-0000000000>psommer",
                    },
                    "title": Object {
                      "fullyHighlighted": false,
                      "matchLevel": "full",
                      "matchedWords": Array [
                        "al",
                      ],
                      "value": "<ais-highlight-0000000000>Al</ais-highlight-0000000000>p-Pflege",
                    },
                  },
                  "brief_description": "Bereich Kulturlandschaft, Biodiversität und Landschaftsqualität",
                  "canton_display_name": "Obwalden",
                  "category_display_name": "Landwirtschaft",
                  "contact_information": "Dieser Betrieb ist noch nicht bei MyZivi registriert. Bitte bewerbe Dich im EZIVI.",
                  "description": "MyDescription <b>Using HTML</b>",
                  "featured_as_new": true,
                  "icon_url": "/myzivi-logo.jpg",
                  "identification_number": 89372,
                  "language": "german",
                  "link": "http://localhost:3000/job_postings/3",
                  "minimum_service_months": 1,
                  "objectID": "3",
                  "organization_display_name": "Alpbetrieb Untersteiglen-Schwand im Tal-Melchsee-Frutt",
                  "plain_description": "100% Dauer Alpsommer: von 01.05. bis 15.10",
                  "preferred_skills": null,
                  "priority_program": true,
                  "publication_date": "2021-06-14",
                  "published": true,
                  "relevancy": 1,
                  "required_skills": "<p> Motivation und physische Kondition für harte körperliche Arbeit im Freien</p>",
                  "sub_category_display_name": null,
                  "title": "Alp-Pflege",
                }
              }
            >
              <JobPosting
                hit={
                  Object {
                    "__position": 1,
                    "_highlightResult": Object {
                      "organization_display_name": Object {
                        "fullyHighlighted": false,
                        "matchLevel": "full",
                        "matchedWords": Array [
                          "al",
                        ],
                        "value": "<ais-highlight-0000000000>Al</ais-highlight-0000000000>",
                      },
                      "plain_description": Object {
                        "fullyHighlighted": false,
                        "matchLevel": "full",
                        "matchedWords": Array [
                          "al",
                        ],
                        "value": "100% Dauer <ais-highlight-0000000000>Al</ais-highlight-0000000000>psommer",
                      },
                      "title": Object {
                        "fullyHighlighted": false,
                        "matchLevel": "full",
                        "matchedWords": Array [
                          "al",
                        ],
                        "value": "<ais-highlight-0000000000>Al</ais-highlight-0000000000>p-Pflege",
                      },
                    },
                    "brief_description": "Bereich Kulturlandschaft, Biodiversität und Landschaftsqualität",
                    "canton_display_name": "Obwalden",
                    "category_display_name": "Landwirtschaft",
                    "contact_information": "Dieser Betrieb ist noch nicht bei MyZivi registriert. Bitte bewerbe Dich im EZIVI.",
                    "description": "MyDescription <b>Using HTML</b>",
                    "featured_as_new": true,
                    "icon_url": "/myzivi-logo.jpg",
                    "identification_number": 89372,
                    "language": "german",
                    "link": "http://localhost:3000/job_postings/3",
                    "minimum_service_months": 1,
                    "objectID": "3",
                    "organization_display_name": "Alpbetrieb Untersteiglen-Schwand im Tal-Melchsee-Frutt",
                    "plain_description": "100% Dauer Alpsommer: von 01.05. bis 15.10",
                    "preferred_skills": null,
                    "priority_program": true,
                    "publication_date": "2021-06-14",
                    "published": true,
                    "relevancy": 1,
                    "required_skills": "<p> Motivation und physische Kondition für harte körperliche Arbeit im Freien</p>",
                    "sub_category_display_name": null,
                    "title": "Alp-Pflege",
                  }
                }
                insights={[Function]}
              >
                <a
                  href="http://localhost:3000/job_postings/3"
                  onClick={[Function]}
                >
                  <div
                    className="card h-100 job-posting-card"
                  >
                    <Ribbon
                      hit={
                        Object {
                          "__position": 1,
                          "_highlightResult": Object {
                            "organization_display_name": Object {
                              "fullyHighlighted": false,
                              "matchLevel": "full",
                              "matchedWords": Array [
                                "al",
                              ],
                              "value": "<ais-highlight-0000000000>Al</ais-highlight-0000000000>",
                            },
                            "plain_description": Object {
                              "fullyHighlighted": false,
                              "matchLevel": "full",
                              "matchedWords": Array [
                                "al",
                              ],
                              "value": "100% Dauer <ais-highlight-0000000000>Al</ais-highlight-0000000000>psommer",
                            },
                            "title": Object {
                              "fullyHighlighted": false,
                              "matchLevel": "full",
                              "matchedWords": Array [
                                "al",
                              ],
                              "value": "<ais-highlight-0000000000>Al</ais-highlight-0000000000>p-Pflege",
                            },
                          },
                          "brief_description": "Bereich Kulturlandschaft, Biodiversität und Landschaftsqualität",
                          "canton_display_name": "Obwalden",
                          "category_display_name": "Landwirtschaft",
                          "contact_information": "Dieser Betrieb ist noch nicht bei MyZivi registriert. Bitte bewerbe Dich im EZIVI.",
                          "description": "MyDescription <b>Using HTML</b>",
                          "featured_as_new": true,
                          "icon_url": "/myzivi-logo.jpg",
                          "identification_number": 89372,
                          "language": "german",
                          "link": "http://localhost:3000/job_postings/3",
                          "minimum_service_months": 1,
                          "objectID": "3",
                          "organization_display_name": "Alpbetrieb Untersteiglen-Schwand im Tal-Melchsee-Frutt",
                          "plain_description": "100% Dauer Alpsommer: von 01.05. bis 15.10",
                          "preferred_skills": null,
                          "priority_program": true,
                          "publication_date": "2021-06-14",
                          "published": true,
                          "relevancy": 1,
                          "required_skills": "<p> Motivation und physische Kondition für harte körperliche Arbeit im Freien</p>",
                          "sub_category_display_name": null,
                          "title": "Alp-Pflege",
                        }
                      }
                    >
                      <div
                        className="featured-ribbon blue"
                      >
                        Neu
                      </div>
                    </Ribbon>
                    <div
                      className="card-body"
                    >
                      <div
                        className="d-flex flex-column"
                      >
                        <div
                          className="icon-container"
                        >
                          <Memo()
                            alt="Alpbetrieb Untersteiglen-Schwand im Tal-Melchsee-Frutt"
                            iconUrl="/myzivi-logo.jpg"
                            key="/myzivi-logo.jpg"
                          >
                            <Component
                              alt="Alpbetrieb Untersteiglen-Schwand im Tal-Melchsee-Frutt"
                              iconUrl="/myzivi-logo.jpg"
                            >
                              <img
                                alt="Alpbetrieb Untersteiglen-Schwand im Tal-Melchsee-Frutt"
                                className="job-posting-icon"
                                src="/myzivi-logo.jpg"
                              />
                            </Component>
                          </Memo()>
                          <h6
                            className="job-posting-subtitle"
                          >
                            Alpbetrieb Untersteiglen-Schwand im Tal-Melchsee-Frutt
                          </h6>
                        </div>
                        <div
                          className="job-posting-card-content"
                        >
                          <h5
                            className="job-posting-title"
                          >
                            Alp-Pflege
                          </h5>
                          <div
                            className="job-description"
                          >
                            Bereich Kulturlandschaft, Biodiversität und Landschaftsqualität
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </a>
              </JobPosting>
            </AlgoliaInsights(JobPosting)>
          </Consumer>
        </Consumer>
      </ConnectorWrapper>
    </div>
  </div>
</Hits>
`;

exports[`CustomHitComponent when no results have been loaded yet renders no results page 1`] = `
<Hits
  createURL={[Function]}
  hasMore={false}
  hasPrevious={false}
  hits={Array []}
  refine={[Function]}
  refineNext={[Function]}
  refinePrevious={[Function]}
>
  <div
    className="row"
  >
    <NoResults>
      <div
        className="d-flex flex-column align-items-center w-100 mt-4"
      >
        <img
          alt="string"
          className="w-25"
          src="test-file-stub"
        />
        <h3
          className="text-muted mt-4"
        >
          string
        </h3>
      </div>
    </NoResults>
  </div>
</Hits>
`;
